### PR TITLE
[Test] Enable numerical test and NamedTuple (for expected) in assert_expected_wrapper

### PR DESCRIPTION
**Pull request by langong347**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #264
* #257
* #263
* #259
* __->__ #262
* #258
* #256
* #255
* #254

Summary:
- Add support for numerical assertion 
- Add additional criteria for differentiating `tuple` from `namedtuple`
- Add `rtol`, `atol` configurability
- Can take both `Dict` and `NamedTuple` for actual and expected

Test Plan:
```
>>>from collections import namedtuple

>>> Out = namedtuple("Out", "x y")
>>> InnerOut = namedtuple("InnerOut", "z")

>>> actual = Out(x=InnerOut(z=tensor([1])), y=tensor([2]))
>>> expected = Out(x=InnerOut(z=tensor([1])), y=tensor([2]))

>>> assert_expected_wrapper(actual, expected)
```

```
$ python -m pytest test/models/test_gpt.py  -vv
================================================= test session starts ==================================================
platform darwin -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /Users/langong/local/miniconda3/envs/t2v/bin/python
cachedir: .pytest_cache
rootdir: /Users/langong/gpt_attention, configfile: pyproject.toml
plugins: mock-3.8.2, cov-3.0.0
collected 18 items

test/models/test_gpt.py::TestMultimodalGPT::test_bad_tokenizers PASSED                                           [  5%]
test/models/test_gpt.py::TestMultimodalGPT::test_fwd_for_generation PASSED                                       [ 11%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward PASSED                                                  [ 16%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_bad_input PASSED                                 [ 22%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_in_modality PASSED                       [ 27%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_out_modality PASSED                      [ 33%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_two_modality PASSED                      [ 38%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_eval_right_shift_on PASSED               [ 44%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_eval_right_shift_off PASSED              [ 50%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_bad_pos_ids PASSED                               [ 55%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_optional_pos_ids PASSED                          [ 61%]
test/models/test_gpt.py::TestTransformerDecoder::test_forward PASSED                                             [ 66%]
test/models/test_gpt.py::TestTransformerDecoder::test_forward_additional_output PASSED                           [ 72%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward PASSED                                        [ 77%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward_masked PASSED                                 [ 83%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward_additional_output PASSED                      [ 88%]
test/models/test_gpt.py::test_sigmoid_linear_unit PASSED                                                         [ 94%]
test/models/test_gpt.py::test_right_shift PASSED                                                                 [100%]

================================================== 18 passed in 1.42s ========
```

Differential Revision: [D38642045](https://our.internmc.facebook.com/intern/diff/D38642045)

### CODE REVIEW

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642045).

File: test/test_utils.py [hunk 1]
```diff
@@ -101,7 +106,12 @@ def is_named_tuple(nt):
     return all(type(name) == str for name in f)
 
 
-def assert_expected_wrapper(actual, expected):
+def assert_expected_wrapper(
```
Comment (RdoubleA) [inline test/test_utils.py:109]: can we name this slightly more descriptive? maybe `assert_expected_named_tuple` or similar?

File: test/test_utils.py [hunk 2]
```diff
@@ -125,18 +135,31 @@ def assert_expected_wrapper(actual, expected):
...
+                _expected_shape, _expected_sum = _expected
+                assert_expected(_actual.shape, _expected_shape)
+                assert_expected(
+                    _actual.sum().item(), _expected_sum, rtol=rtol, atol=atol
+                )
+            elif isinstance(_expected, Tensor):
+                # test value
+                assert_expected(_actual, _expected, rtol=rtol, atol=atol)
+            else:
+                raise TypeError(
```
Comment (RdoubleA) [inline test/test_utils.py:160]: can we print the type of _expected in the error message?

**Review by RdoubleA — APPROVED** [review]

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642045).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642045).

**Pull request closed without merging**